### PR TITLE
🐙 octavia-cli: use model serialization

### LINKED ISSUE #12270: 🐛 octavia-cli: connections re-created on apply
## Current Behavior
When a user generates a connection configuration and edits fields such as `namespace_definition` or `namespace_format` the connection is created but subsequent `octavia apply` recreates the same connection again leading to multiple connections getting created for a single config.

## Expected Behavior
The original connection should be updated on configuration changes.

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Some shortcuts taken in the early development of the project challenge the stability of the CLI and lead to issues such as #12270 

## Problems and solutions

**Problem: We discarded the use of model serialization**
We disabled model serialization and validation on some API responses because it required updating the Open API config to declare fields as nullable (more about this [here](https://github.com/airbytehq/airbyte/pull/12260)). It means the response we handled were raw python dict and not validated models.
**Solution:** edit [airbyte-api/src/main/openapi/config.yaml](https://github.com/airbytehq/airbyte/pull/12133/files#diff-8d2a55215b7d1dbfdb14f80e5fb444c9b940be967bdfb8cfbf1e24703579095e) to make some field nullable. Do not pass `_check_return_type=False` when calling endpoints to retrieve pure API client models that are validated and expose the convenient model interfaces + python casing. **This allows us**:
* **Use snake case fields on connection** and have consistent casing in generated resources (source / destination / connection). See [octavia-cli/integration_tests/configurations/connections/poke_to_pg/configuration.yaml](https://github.com/airbytehq/airbyte/pull/12133/files#diff-df82ef9c861f439f0ef2cb8652bffee700bdfcd553973c9406ba7bfc0828f1f4)
* **Build valid `ConnectionUpdate` and `ConnectionCreate` payloads** with the `_deserialize_raw_configuration` method in [octavia-cli/octavia_cli/apply/resources.py](https://github.com/airbytehq/airbyte/pull/12133/files#diff-7667f25ac662725c089e581128312f38cd66f25a7a760d4502f660753e6ed123). `BaseResource` instances now have two configuration attributes: `raw_configuration` which is the content of the YAML configuration serialized as a Python dict and `configuration` which is the output of `_deserialize_raw_configuration`. By default it just extract the `configuration` key of `raw_configuration`, but for `Connection` this methods builds a `configuration` dict with value corresponding to API Client models for validation.

**Problem: We used the `search` endpoint to retrieve remote resources**
The `search` endpoint for connection retrieved inconsistent results when a connection was edited (which led to #12270). We used the `search` API endpoint in a early stage of the project when we did not persist the  `resource_id` in the `state` file.
**Solution:** Now that we have `state` file we can use the `get` endpoints to retrieve remote resource from `resource_id`stored in the `state`. These change happen in [octavia-cli/octavia_cli/apply/resources.py](https://github.com/airbytehq/airbyte/pull/12133/files#diff-7667f25ac662725c089e581128312f38cd66f25a7a760d4502f660753e6ed123)

## Other changes
I introduced additional changes to make the code lighter and leverage inheritance as much as possible: I created a `SourceAndDestination` from `BaseResource` because `Source` and `Destination` share duplicate code otherwise.

## 🚨 User Impact 🚨
**Connection configurations** are now using `snake_case` instead of `camel_case`. Users with existing  connection configuration will have to edit their `configuration.yaml` file for connections to change the casing.

### CODE REVIEW

File: octavia-cli/README.md [hunk 1]
```diff
@@ -362,8 +362,9 @@ You can disable telemetry by setting the `OCTAVIA_ENABLE_TELEMETRY` environment
...
 ## Changelog
 
-| Version | Date       | Description       | PR                                                       |
-|---------|------------|-------------------|----------------------------------------------------------|
-| 0.36.2  | 2022-04-15 | Improve telemetry | [#12072](https://github.com/airbytehq/airbyte/issues/11896)|
-| 0.35.68 | 2022-04-12 | Add telemetry     | [#11896](https://github.com/airbytehq/airbyte/issues/11896)|
-| 0.35.61 | 2022-04-07 | Alpha release     | [EPIC](https://github.com/airbytehq/airbyte/issues/10704)|
+| Version | Date       | Description                         | PR                                                       |
+|---------|------------|-------------------------------------|----------------------------------------------------------|
+| 0.36.7  | 2022-05-03 | Use camel case in connection fields | [#12133](https://github.com/airbytehq/airbyte/pull/12133)|
```
Comment (lmossman) [inline octavia-cli/README.md:367]:
```suggestion
| 0.36.7  | 2022-05-03 | Use snake case in connection fields | [#12133](https://github.com/airbytehq/airbyte/pull/12133)|
```

File: octavia-cli/integration_tests/configurations/connections/poke_to_pg/configuration.yaml [hunk 2]
```diff
@@ -1,38 +1,35 @@
...
-  name: poke_to_g # OPTIONAL | string | Optional name of the connection
-  namespaceDefinition: source # OPTIONAL | string | Allowed values: source, destination, customformat
-  namespaceFormat: "${SOURCE_NAMESPACE}" # OPTIONAL | string | Used when namespaceDefinition is 'customformat'. If blank then behaves like namespaceDefinition = 'destination'. If "${SOURCE_NAMESPACE}" then behaves like namespaceDefinition = 'source'.
+  namespace_definition: source # OPTIONAL | string | Allowed values: source, destination, customformat
+  namespace_format: "${SOURCE_NAMESPACE}" # OPTIONAL | string | Used when namespaceDefinition is 'customformat'. If blank then behaves like namespaceDefinition = 'destination'. If "${SOURCE_NAMESPACE}" then behaves like namespaceDefinition = 'source'.
   prefix: "" # REQUIRED | Prefix that will be prepended to the name of each stream when it is written to the destination
-  resourceRequirements: # OPTIONAL | object | Resource requirements to run workers (blank for unbounded allocations)
+  resource_requirements: # OPTIONAL | object | Resource requirements to run workers (blank for unbounded allocations)
```
Comment (lmossman) [inline octavia-cli/integration_tests/configurations/connections/poke_to_pg/configuration.yaml:13]: Try to catch api exception or validate existing configs to provide a understandable error message to the user directing them to switch to snake_case, or re-generate the files.

File: octavia-cli/octavia_cli/apply/resources.py [hunk 3]
```diff
@@ -526,35 +525,47 @@ def update_payload(self) -> ConnectionUpdate:
...
-        return {k: v for k, v in comparable_configuration.items() if k not in keys_to_filter_out}
+    def _get_remote_comparable_configuration(self) -> dict:
+        keys_to_filter_out = [
+            "name",
+            "source_id",
+            "destination_id",
+            "connection_id",
+            "operation_ids",
+            "source_catalog_id",
+        ]  # We do not allow local edition of these keys
```
Comment (lmossman) [inline octavia-cli/octavia_cli/apply/resources.py:567]:
```suggestion
        ]  # We do not allow local editing of these keys
```

**Review by lmossman — APPROVED** [review]
Overall looks good me, if you want me to take another look after making the validation of snake_case changes, let me know

**Comment by alafanechere** [comment]
@lmossman I implemented the check for camel case in configuration with this commit https://github.com/airbytehq/airbyte/pull/12133/commits/bf678e07baed53446c3452d1f3217e01c7b69ab2

I won't write a script for users to update their config as I think it does not worth it considering the current adoption of the tool. I'll prefer to suggest a script on our forum if some user complains about this manual operation.

**Pull request merged**